src/Npgsql/Replication/NpgsqlReplicationConnection.cs

```diff
@@ -0,0 +1,888 @@
+            EnsureState(ReplicationConnectionState.Streaming);
+            if (await _feedbackSemaphore.WaitAsync(Timeout.Infinite, cancellationToken))
+                await SendFeedback(cancellationToken: cancellationToken);
+        }
+
+        async Task SendFeedback(bool requestReply = false, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                // Disable the timers while we are sending
+                _sendFeedbackTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+                _requestFeedbackTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+
+                var connector = _npgsqlConnection.Connector!;
+                var buf = connector.WriteBuffer;
+
+                const int len = 39;
+
+                if (buf.WriteSpaceLeft < len)
+                    await connector.Flush(true, cancellationToken);
+
+                buf.WriteByte(FrontendMessageCode.CopyData);
+                buf.WriteInt32(len - 1);
+                buf.WriteByte((byte)'r');  // TODO: enum/const?
+                // We write the LSNs as Int64 here to save us the casting
+                buf.WriteInt64(Interlocked.Read(ref _lastReceivedLsn));
+                buf.WriteInt64(Interlocked.Read(ref _lastFlushedLsn));
+                buf.WriteInt64(Interlocked.Read(ref _lastAppliedLsn));
+                buf.WriteInt64(TimestampHandler.ToPostgresTimestamp(DateTime.Now));
+                buf.WriteByte(requestReply ? (byte)1 : (byte)0);
+
+                await connector.Flush(true, cancellationToken);
+            }
+            finally
+            {
+                // Restart the timers if we are still streaming
+                if (State == ReplicationConnectionState.Streaming)
+                {
+                    _sendFeedbackTimer.Change(WalReceiverStatusInterval, Timeout.InfiniteTimeSpan);
+                    _requestFeedbackTimer.Change(_requestFeedbackInterval, Timeout.InfiniteTimeSpan);
+                }
+                _feedbackSemaphore.Release();
+            }
+        }
+
+        async void TimerRequestFeedback(object? obj)
+        {
+            try
+            {
+                if (State != ReplicationConnectionState.Streaming)
+                    return;
+
+                if (await _feedbackSemaphore.WaitAsync(Timeout.Infinite))
+                    await SendFeedback(true);
+            }
+            catch (Exception e)
+            {
+                Log.Error("An exception occurred while requesting streaming replication feedback from the server.", e, _npgsqlConnection?.Connector?.Id ?? 0);
+            }
+        }
+
+        async void TimerSendFeedback(object? obj)
+        {
+            try
+            {
+                if (State != ReplicationConnectionState.Streaming)
+                    return;
+
+                if (await _feedbackSemaphore.WaitAsync(TimeSpan.Zero))
+                {
+                    await SendFeedback();
```

Comment:
This could run after streaming completed (race condition with streaming completion). In that case, SendFeedback could be called at the same time as some other user operation (e.g. DropReplicationSlot) or closing the connection.

There's two general options here:

1. Make sure that everything that happens takes _feedbackSemaphore, including stuff like IdentifySystem. This is the approach done for regular connection and keepalive - that's why any operation must always start a UserAction.
2. When streaming completes, wait until the timer completes by disposing it, passing a SemaphoreSlim as the  WaitHandle (or just doing DisposeAsync in more modern TFMs). This way we can be sure that when replication completes, there's no more timer. This also removes the need to check the state in TimerSendFeedback (since it only ever runs during replication).

Although up to now we've gone with the 1st option, we should consider doing the 2nd which seems cleaner - the less statefulness, the better...

/cc @vonzshik who is interested in timers and stuff - any thoughts?

Reply:
> and stuff 

:trollface:

Okay, I think I more or less understand that's going on here. 

I think the second option does make sense. Although, the only change would be around dispose - wouldn't it?

But why do we have two almost the same keepalives?

Reply:
> But why do we have two almost the same keepalives?

Yeah, see my main review comment in https://github.com/npgsql/npgsql/pull/3281#pullrequestreview-524121816. Technically, one keepalive is for PostgreSQL (the one where we don't request a reply), and the other is for us (where we ask PostgreSQL to send us a keepalive). So it could be said that both should be treated and configured separately. But I agree that have just the second one probably fulfils user needs (and would reduce complexity).

Reply:
> I think the second option does make sense. Although, the only change would be around dispose - wouldn't it?

I think it means we'd have to dispose the timer(s) at the end of replication. Right now the PR doesn't do that - it disposes only when the entire connection is disposed. Basically it would mean that the timer itself lives only for the duration of the replication, which makes sense if it's only supposed to be active at that point.

Reply:
> This could run after streaming completed

We'd return four lines above if we aren't streaming - no?

Reply:
There's a race condition - streaming could be stopping just after we did that check, no? Especially as _feedbackSemaphore isn't taken at any point when stopping the timer.

Reply:
> Especially as _feedbackSemaphore isn't taken at any point when stopping the timer.

It should be taken in all places where we're stopping the timers.
The most problematic one being Cancel/Dispose because we can't block forever there.

Reply:
https://github.com/Brar/npgsql/pull/10